From a4c2509ed9656f0299011e8bd8fcd0fa0dacdc4d Mon Sep 17 00:00:00 2001
From: Slawek Kaplonski <skaplons@redhat.com>
Date: Wed, 30 Jul 2025 12:06:25 +0200
Subject: [PATCH] Copy alembic migration scripts newer than 2023.1 to 2023.1
 folder

Alembic migration scripts which were introduced in upstream after
the 2023.1 version are copied with this patch to the 2023.1 alembic
migration directory in order to be able to apply them by the
neutron-db-migration scrip which runs in RHOSO 18.0 (based on u/s
2023.1).
This is downstream only workaround to be able to apply those db changes
in RHOSO 18.

Related: #OSPRH-15215
Downstream-Only
---
 ...3_add_pk_firewall_group_associations_v2.py | 53 +++++++++++++
 .../6941ce70131e_add_standard_attr_id.py      | 78 +++++++++++++++++++
 2 files changed, 131 insertions(+)
 create mode 100644 neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/2a0d33e9ef63_add_pk_firewall_group_associations_v2.py
 create mode 100644 neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/6941ce70131e_add_standard_attr_id.py

diff --git a/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/2a0d33e9ef63_add_pk_firewall_group_associations_v2.py b/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/2a0d33e9ef63_add_pk_firewall_group_associations_v2.py
new file mode 100644
index 000000000..b280d9ea5
--- /dev/null
+++ b/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/2a0d33e9ef63_add_pk_firewall_group_associations_v2.py
@@ -0,0 +1,53 @@
+# Copyright 2019 Canonical Ltd.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+"""add pk firewall_group_associations_v2
+
+Revision ID: 2a0d33e9ef63
+Revises: 6941ce70131e
+Create Date: 2025-01-20 18:00:00.000000
+
+"""
+
+from alembic import op
+from sqlalchemy.engine import reflection
+
+from oslo_log import log as logging
+
+
+# revision identifiers, used by Alembic.
+revision = '2a0d33e9ef63'
+down_revision = '6941ce70131e'
+
+LOG = logging.getLogger(__name__)
+
+
+def upgrade():
+    bind = op.get_bind()
+    insp = reflection.Inspector.from_engine(bind.engine)
+    if 'firewall_group_port_associations_v2' not in insp.get_table_names():
+        return
+    pk = insp.get_pk_constraint('firewall_group_port_associations_v2')
+    if not pk['constrained_columns']:
+        op.create_primary_key(
+            'pk_firewall_group_port_associations_v2',
+            'firewall_group_port_associations_v2',
+            ['firewall_group_id', 'port_id'])
+    else:
+        # Revision '6941ce70131e' has been updated to create the
+        # missing PK. Depending whether the env is already deployed or
+        # not we may or not have to add the primary key.
+        LOG.info("The primary key in firewall_group_port_associations_v2 "
+                 "already exists, continuing.")
diff --git a/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/6941ce70131e_add_standard_attr_id.py b/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/6941ce70131e_add_standard_attr_id.py
new file mode 100644
index 000000000..27213be89
--- /dev/null
+++ b/neutron_fwaas/db/migration/alembic_migrations/versions/2023.1/expand/6941ce70131e_add_standard_attr_id.py
@@ -0,0 +1,78 @@
+# Copyright 2023 EasyStack Limited
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+"""add standard attributes
+
+Revision ID: 6941ce70131e
+Revises: f24e0d5e5bff
+Create Date: 2022-12-01 04:19:57.324584
+
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+# revision identifiers, used by Alembic.
+revision = '6941ce70131e'
+down_revision = 'f24e0d5e5bff'
+tables = ['firewall_groups_v2', 'firewall_rules_v2', 'firewall_policies_v2']
+
+
+standardattrs = sa.Table(
+    'standardattributes', sa.MetaData(),
+    sa.Column('id', sa.BigInteger(), primary_key=True, autoincrement=True),
+    sa.Column('resource_type', sa.String(length=255), nullable=False),
+    sa.Column('description', sa.String(length=255), nullable=True))
+
+
+def generate_records_for_existing(table):
+    model = sa.Table(table, sa.MetaData(),
+                     sa.Column('id', sa.String(length=36), nullable=False),
+                     sa.Column('description', sa.String(length=255),
+                               nullable=True),
+                     sa.Column('standard_attr_id', sa.BigInteger(),
+                               nullable=True))
+    session = sa.orm.Session(bind=op.get_bind())
+    with session.begin(subtransactions=True):
+        for row in session.query(model):
+            res = session.execute(
+                standardattrs.insert().values(resource_type=table,
+                                              description=row[1])
+            )
+            session.execute(
+                model.update().values(
+                    standard_attr_id=res.inserted_primary_key[0]).where(
+                        model.c.id == row[0])
+            )
+    session.commit()
+
+
+def upgrade():
+    for table in tables:
+        op.add_column(table, sa.Column('standard_attr_id', sa.BigInteger(),
+                                       nullable=True))
+        op.create_foreign_key(
+            constraint_name=None, source_table=table,
+            referent_table='standardattributes',
+            local_cols=['standard_attr_id'], remote_cols=['id'],
+            ondelete='CASCADE')
+        generate_records_for_existing(table)
+        op.alter_column(table, 'standard_attr_id', nullable=False,
+                        existing_type=sa.BigInteger(), existing_nullable=True,
+                        existing_server_default=False)
+        op.create_unique_constraint(
+            constraint_name='uniq_%s0standard_attr_id' % table,
+            table_name=table, columns=['standard_attr_id'])
+        op.drop_column(table, 'description')
-- 
2.50.1

